src/target/ext/host_io.rs

```diff
@@ -0,0 +1,228 @@
+        /// Exclusive access.
+        const O_EXCL = 0x800;
+    }
+}
+
+bitflags! {
+    /// Host file permissions.
+    /// [mode_t Values]: https://sourceware.org/gdb/onlinedocs/gdb/mode_005ft-Values.html
+    pub struct HostMode: u32 {
+        /// A regular file.
+        const S_IFREG = 0o100000;
+        /// A directory.
+        const S_IFDIR = 0o40000;
+        /// User read permissions.
+        const S_IRUSR = 0o400;
+        /// User write permissions.
+        const S_IWUSR = 0o200;
+        /// User execute permissions.
+        const S_IXUSR = 0o100;
+        /// Group read permissions.
+        const S_IRGRP = 0o40;
+        /// Group write permissions
+        const S_IWGRP = 0o20;
+        /// Group execute permissions.
+        const S_IXGRP = 0o10;
+        /// World read permissions.
+        const S_IROTH = 0o4;
+        /// World write permissions
+        const S_IWOTH = 0o2;
+        /// World execute permissions.
+        const S_IXOTH = 0o1;
+    }
+}
+
+/// An interface to send pread data back to the GDB client.
+pub struct HostIoOutput<'a> {
+    cb: &'a mut dyn FnMut(&[u8]),
+}
+
+impl<'a> HostIoOutput<'a> {
+    pub(crate) fn new(cb: &'a mut dyn FnMut(&[u8])) -> Self {
+        Self { cb }
+    }
+
+    /// Write out raw file bytes to the GDB debugger.
+    pub fn write(self, buf: &[u8]) {
+        (self.cb)(buf)
+    }
+}
+
+/// Target Extension - Perform I/O operations on host
+pub trait HostIo: Target {
+    /// Enable open operation.
+    #[inline(always)]
+    fn enable_open(&mut self) -> Option<HostIoOpenOps<Self>> {
+        None
+    }
+    /// Enable close operation.
+    #[inline(always)]
+    fn enable_close(&mut self) -> Option<HostIoCloseOps<Self>> {
+        None
+    }
+    /// Enable pread operation.
+    #[inline(always)]
+    fn enable_pread(&mut self) -> Option<HostIoPreadOps<Self>> {
+        None
+    }
+    /// Enable pwrite operation.
+    #[inline(always)]
+    fn enable_pwrite(&mut self) -> Option<HostIoPwriteOps<Self>> {
+        None
+    }
+    /// Enable fstat operation.
+    #[inline(always)]
+    fn enable_fstat(&mut self) -> Option<HostIoFstatOps<Self>> {
+        None
+    }
+    /// Enable unlink operation.
+    #[inline(always)]
+    fn enable_unlink(&mut self) -> Option<HostIoUnlinkOps<Self>> {
+        None
+    }
+    /// Enable readlink operation.
+    #[inline(always)]
+    fn enable_readlink(&mut self) -> Option<HostIoReadlinkOps<Self>> {
+        None
+    }
+    /// Enable setfs operation.
+    #[inline(always)]
+    fn enable_setfs(&mut self) -> Option<HostIoSetfsOps<Self>> {
+        None
+    }
+}
+
+define_ext!(HostIoOps, HostIo);
+
+/// Nested Target Extension - Host I/O open operation.
+pub trait HostIoOpen: HostIo {
+    /// Close the open file corresponding to fd and return 0, or -1 if an error
+    /// occurs.
+    fn open(
+        &mut self,
+        filename: &[u8],
+        flags: HostOpenFlags,
+        mode: HostMode,
+    ) -> TargetResult<i32, Self>;
```

Comment:
`TargetResult` is definitely not the right type to use. The `TargetResult` type is used on methods that support returning `Exx` as a error code. The `vFile` packets have their own error reporting mechanism, hence why I suggested the `HostIoResult` + `HostIoError` type.